Add Linux/Unix support

TheComputerGuy96 opened the pull request:
This is currently a draft until I can get SilentPatch actually compiling/working with MinGW GCC (I got the ddraw component to compile though)

TheComputerGuy96 committed:
Use lowercase include path for windows.h

MinGW GCC only ships the lower-case version of that file (which
causes issues on Unix environments)

TheComputerGuy96 committed:
Patterns: Add template definition in get_first()

MinGW GCC complains without it being present

TheComputerGuy96 committed:
MemoryMgr: Actually return in DynBase::ReadCallFrom()

MinGW GCC warns about this oversight

TheComputerGuy96 committed:
Patterns: Add missing cstdint include

MinGW GCC needs it

TheComputerGuy96 committed:
Patterns: Use GCC/Clang-friendly define for detecting exception support

This fixes a compile error when compiling the SilentPatch DDraw component
with MinGW GCC

TheComputerGuy96 committed:
MemoryMgr: Add GCC/Clang-specific assembly statements

Both of those compilers don't support the MSVC inline assembly syntax

TheComputerGuy96 committed:
Patterns: Don't redefine the NOMINMAX macro

MinGW GCC defines it by default (and that causes a compile warning)

CookiePLMonster commented:
This looks very promising, looking forward to that!

TheComputerGuy96 commented:
This PR should be okay now (because I tested it with SilentPatchIII/VC that's locally compiled with MinGW GCC and both of those libraries compile/work fine)

CookiePLMonster commented:
> This PR should be okay now (because I tested it with SilentPatchIII/VC that's locally compiled with MinGW GCC and both of those libraries compile/work fine)

Good stuff! I'll check on Windows locally with SP for II/VC (using a VS2017 toolset) and with any other patch using VS2022, if everything works good then I'll merge. I haven't spotted anything wrong in the review though so it should be all good.

TheComputerGuy96 committed:
MemoryMgr.GTA: Switch to signed integers in AddrVariant

MinGW GCC prefers signed integers for integer constants over unsigned ones
which causes conversion errors for those cases

TheComputerGuy96 committed:
MemoryMgr: Don't set NOVMT for non-MSVC-compatible compilers

The novtable attribute isn't supported by MinGW GCC (which causes
a compile warning)

CookiePLMonster commented:
I see you're still making changes, is this ready to go or should I wait?

TheComputerGuy96 commented:
> I see you're still making changes, is this ready to go or should I wait?

The III/VC parts seem pretty stable (SA has weird segfault issues though) so this is good enough

CookiePLMonster commented on MemoryMgr.GTA.h:
```diff
@@ -20,7 +20,7 @@ namespace Memory
 	};
 
-	using AddrVariant = std::variant<uintptr_t, PatternAndOffset>;
+	using AddrVariant = std::variant<intptr_t, uintptr_t, PatternAndOffset>;
```
I wonder if `uintptr_t` is needed here to begin with. If it's not, it'd simplify the `get_if` accesses quite a lot.

TheComputerGuy96 replied:
I get errors like this without allowing `intptr_t` in AddrVariant:
```
/mnt/storagebeast/Development/General/SilentPatch/SilentPatchSA/AudioHardwareSA.cpp: At global scope:
/mnt/storagebeast/Development/General/SilentPatch/SilentPatchSA/AudioHardwareSA.cpp:6:65: error: no matching function for call to ‘AddressByVersion<void*>(int, int, int)’
    6 | static void* CAEDataStream__Initialise = AddressByVersion<void*>(0x4DC2B0, 0x4DC7A0, 0x4E7550);
      |                                          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/storagebeast/Development/General/SilentPatch/SilentPatch/Utils/MemoryMgr.GTA.h:385:10: note: candidate: ‘T AddressByVersion(Memory::AddrVariant, Memory::AddrVariant, Memory::AddrVariant) [with T = void*; Memory::AddrVariant = std::variant<unsigned int, Memory::PatternAndOffset>]’
  385 | inline T AddressByVersion(Memory::AddrVariant address10, Memory::AddrVariant address11, Memory::AddrVariant addressSteam)
      |          ^~~~~~~~~~~~~~~~
/mnt/storagebeast/Development/General/SilentPatch/SilentPatch/Utils/MemoryMgr.GTA.h:385:47: note:   no known conversion for argument 1 from ‘int’ to ‘Memory::AddrVariant’ {aka ‘std::variant<unsigned int, Memory::PatternAndOffset>’}
  385 | inline T AddressByVersion(Memory::AddrVariant address10, Memory::AddrVariant address11, Memory::AddrVariant addressSteam)
```

(I guess I could cast all of those constants but that would add extra complexity)

CookiePLMonster replied:
That's the signed one, and I'm wondering if there's any point in keeping the unsigned variant.

TheComputerGuy96 replied:
Removing the unsigned variant does work with MinGW GCC (but MSVC might be different)

CookiePLMonster replied:
I will verify that locally on msvc later. This functionality only exists for GTA so there is no risk of breaking other games.

CookiePLMonster commented on MemoryMgr.GTA.h:
```diff
@@ -268,7 +268,9 @@ namespace Memory
-					const uintptr_t addr = *std::get_if<uintptr_t>(&address10);
+					const uintptr_t addr = (std::get_if<uintptr_t>(&address10) != nullptr)
+											? *std::get_if<uintptr_t>(&address10)
+											: std::make_unsigned_t<intptr_t>(*std::get_if<intptr_t>(&address10));
```
That might require `<type_traits>` to be included I think.

TheComputerGuy96 replied:
Done

CookiePLMonster commented on MemoryMgr.h:
```diff
@@ -6,17 +6,28 @@
 #define DEPRECATED __declspec(deprecated)
+#define WRAPARG(a) ((int)a)
+
+#ifdef _MSVC_VER
```
It's `_MSC_VER`.

TheComputerGuy96 replied:
That was a fairly easy definition to confuse (anyway I changed that)

CookiePLMonster commented on MemoryMgr.h:
```diff
@@ -6,17 +6,28 @@
+#define VARJMP(a) { __asm__ volatile("jmp %0" :: "m" (a)); }
+#endif
 
+#ifdef _MSVC_VER
```
Ditto.

TheComputerGuy96 replied:
Done

CookiePLMonster reviewed (changes requested):
MSVC check has to be changed, looks good otherwise.

With this one change done locally, I tested the following:
* **SilentPatch** (GTA) - v141_xp toolset, 32-bit, works fine.
* **SilentPatchCMR3** - v143 toolset, 32-bit, works fine.
* **SilentPatchYRC** (with fixes for the latest ModUtils) - v143 toolset, 64-bit, works fine.

Once this one issue is fixed, I will merge.

CookiePLMonster commented:
All good now, thank you!

EDIT:
One more question, were you not getting conversion warnings on `AddressByVersion` in III and VC that use `uintptr_t address10, uintptr_t address11, uintptr_t addressSteam` directly? I'm thinking if I should turn all those `uintptr_t`'s into `intptr_t`'s in a follow-up change. The results are being typecasted anyway and only the internals are operating on integers.

TheComputerGuy96 commented:
I just realized `std::basic_string<unsigned char>` (or `uint8_t`) isn't part of the C++ standard and so recent llvm-mingw can't compile the Patterns stuff that's required for all of SilentPatch (thankfully C++20 introduces a new `char8_t` type which is basically identical to those previous types but there's the last 1% missing so I have to do some type hackery)

I'll have to submit another MR to work around that (and potentially fix another issue)

CookiePLMonster commented:
The lowest common denominator for ModUtils is C++17, so if possible try to make it work there. It's weird though, I thought `basic_string` can take any arbitrary type?

TheComputerGuy96 commented:
> I thought basic_string can take any arbitrary type?

The arbitrary type support is compiler-specific (here are the standardized types: https://en.cppreference.com/w/cpp/string/char_traits)

CookiePLMonster commented:
Would defining custom traits for uint8_t and making these `basic_string` use it be an option? It could probably mostly inherit from `char_traits<char>`.

EDIT:
Seems pretty easy: https://github.com/nlohmann/json/blob/develop/include/nlohmann/detail/meta/type_traits.hpp#L194-L216

So if you go for private `pattern_traits` for those patterns, I will accept the PR.
